Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
+// **************
+
+#[allow(dead_code)]
+struct Map<K, V> 
+  where K: Clone + Eq + std::hash::Hash, V: Clone
+{
+    data: Rc<Sequence<(K, V)>>,
+    // Any time we explicitly access this map, we index the data
+    cache: RefCell<Option<HashMap<K, V>>>
+}
+
+#[allow(dead_code)]
+impl <K, V> Map<K, V>
+  where K: Clone + Eq + std::hash::Hash, V: Clone
+{
+    fn new_empty() -> Rc<Map<K, V>> {
+        Rc::new(Map {
+            data: Sequence::new_array_sequence(&Rc::new(Vec::new())),
+            cache: RefCell::new(None)
+        })
+    }
+    fn new_from_sequence(data: &Rc<Sequence<(K, V)>>) -> Rc<Map<K, V>> {
+        Rc::new(Map {
+            data: Rc::clone(data),
+            cache: RefCell::new(None)
+        })
+    }
+    fn to_hashmap_owned<K2, V2>(&self, converter_k: fn(&K)->K2, converter_v: fn(&V)->V2) -> HashMap<K2, V2>
+      where K2: Eq + std::hash::Hash, V2: Clone
+    {
+        let mut result: HashMap<K2, V2> = HashMap::new();
+        for (k, v) in self.data.to_array().iter() {
+            result.insert(converter_k(k), converter_v(v));
+        }
+        result
+    }
+    fn compute_hashmap(&self) {
+        let mut cache = self.cache.borrow_mut();
+        if cache.is_none() {
+            *cache = Some(self.to_hashmap_owned(
+                |x| x.clone(),
+                |x| x.clone()
+            ));
+        }
+    }
+    // Dafny will normally guarantee that the key exists.
+    fn get(&self, key: &K) -> V {
+        self.compute_hashmap();
+        self.cache.borrow_mut().as_ref().unwrap().get(key).unwrap().clone()
+    }
+    fn add(&self, key: K, value: V) -> Rc<Map<K, V>> {
+        let new_data = Sequence::<(K, V)>::new_array_sequence(
+            &Rc::new(vec![(key, value)]));
+        let combined_data = Sequence::<(K, V)>::new_concat_sequence(
+            &self.data, &new_data);
+        Rc::new(Map {
+            data: combined_data,
+            cache: RefCell::new(None)
+        })
+    }
+    fn add_multiple(&self, other: &Rc<Map<K, V>>) -> Rc<Map<K, V>> {
+        let new_data = Rc::clone(&other.data);
+        let combined_data = Sequence::<(K, V)>::new_concat_sequence(
+            &self.data, &new_data);
+        Rc::new(Map {
+            data: combined_data,
+            cache: RefCell::new(None)
+        })
+    
+    }
+    // TODO: Remaining methods + check names
+}
+
+// Generic function to allocate and return a raw pointer immediately
+#[inline]
+pub fn allocate<T>(value: Box<T>) -> *const T {
+    Box::into_raw(value)
+}
+
+// Generic function to safely deallocate a raw pointer
+#[inline]
+pub fn deallocate<T : ?Sized>(pointer: *const T) {
+    unsafe {
+        // Takes ownership of the reference,
+        // so that it's deallocated at the end of the method
+        let _ = Box::from_raw(pointer as *mut T);
+    }
+}
+
+pub fn is_instance_of<T: ?Sized + Any, U: 'static>(theobject: *const T) -> bool {
+    unsafe { &*theobject }.as_any().downcast_ref::<U>().is_some()
```

Comment:
Because Dafny is pretty restrictive about the types on either side of an `is`, it seems like it's possible to not resort to `as_any()`, as much as that is still correct. How much of an efficiency hit is it?

Reply:
Based on my current understanding, Rust is one of this few languages where references are not "transparent". Transparency means that the representation of the reference is the same whether the object in the original code was a class or an interface.
Contrary to Java, that means if a struct `C` implement `trait A` and `trait B`, a pointer of type `*const C` does not have the same representation as a pointer of type `*const dyn A` or of type `*const dyn B`. This is because a `dyn` pointer not only stores the address of the object data, but also the address of the object's methods (that's why there is a keyword "dyn", to indicate it stores two pointers). A pointer of type `*const C` therefore is not immediately a pointer of type `*const dyn A`, and a pointer of type `*const dyn A` requires an explicit conversion to a pointer of type `*const dyn B` because the methods are not at the same address. This is why the method `as_any()` makes it possible to go back on our feet.

By the way, I'm not using this is_instance_of yet but I will shortly not in the next PR, but in a PR afterwards. So this can still evolve.

Reply:
Yeah, no reflection is needed because any class that can be upcasted must implement the Any trait, otherwise the approach is unsound.